src/backend/base/langflow/services/tracing/service.py

```diff
@@ -46,232 +47,346 @@ def _get_arize_phoenix_tracer():
+            trace_context_var.set(trace_context)
+            await self._start(trace_context)
+            self._initialize_langsmith_tracer(trace_context)
+            self._initialize_langwatch_tracer(trace_context)
+            self._initialize_langfuse_tracer(trace_context)
+            self._initialize_arize_phoenix_tracer(trace_context)
+        except Exception as e:  # noqa: BLE001
+            logger.debug(f"Error initializing tracers: {e}")
+
+    async def _stop(self, trace_context: TraceContext) -> None:
+        try:
+            trace_context.running = False
+            # check the qeue is empty
+            if not trace_context.traces_queue.empty():
+                await trace_context.traces_queue.join()
+            if trace_context.worker_task:
+                trace_context.worker_task.cancel()
+                trace_context.worker_task = None
+
+        except Exception:  # noqa: BLE001
+            logger.exception("Error stopping tracing service")
+
+    def _end_all_tracers(self, trace_context: TraceContext, outputs: dict, error: Exception | None = None) -> None:
+        for tracer in trace_context.tracers.values():
+            if tracer.ready:
+                try:
+                    # why all_inputs and all_outputs? why metadata=outputs?
+                    tracer.end(
+                        trace_context.all_inputs,
+                        outputs=trace_context.all_outputs,
+                        error=error,
+                        metadata=outputs,
+                    )
+                except Exception:  # noqa: BLE001
+                    logger.exception("Error ending all traces")
+
+    async def end_tracers(self, outputs: dict, error: Exception | None = None) -> None:
+        """End the trace for a graph run.
+
+        - stop worker for current trace_context
+        - call end for all the tracers
+        """
+        if self.deactivated:
+            return
+        trace_context = trace_context_var.get()
+        if trace_context is None:
+            msg = "called end_tracers but no trace context found"
+            raise RuntimeError(msg)
+        await self._stop(trace_context)
+        self._end_all_tracers(trace_context, outputs, error)
 
-    def set_run_id(self, run_id: UUID) -> None:
-        self.run_id = run_id
+    @staticmethod
+    def _cleanup_inputs(inputs: dict[str, Any]):
+        inputs = inputs.copy()
+        for key in inputs:
+            if "api_key" in key:
+                inputs[key] = "*****"  # avoid logging api_keys for security reasons
+        return inputs
 
-    def _start_traces(
+    def _start_component_traces(
         self,
-        trace_id: str,
-        trace_name: str,
-        trace_type: str,
-        inputs: dict[str, Any],
-        metadata: dict[str, Any] | None = None,
-        vertex: Vertex | None = None,
+        ctc: ComponentTraceContext,
+        ct: TraceContext,
```

Comment:
would prefer explicit names instead of abbreviations.
```suggestion
        component_trace_context: ComponentTraceContext,
        trace_context: TraceContext,
```